owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/security/PassCodeViewModel.kt

```diff
@@ -64,6 +83,40 @@ class PassCodeViewModel(
     fun setMigrationRequired(required: Boolean) =
         preferencesProvider.putBoolean(PassCodeActivity.PREFERENCE_MIGRATION_REQUIRED, required)
 
+    fun setLastUnlockTimestamp() =
+        preferencesProvider.putLong(PREFERENCE_LAST_UNLOCK_TIMESTAMP, SystemClock.elapsedRealtime())
+
+    fun getNumberOfAttempts() = preferencesProvider.getInt(PREFERENCE_LOCK_ATTEMPTS, 0)
+
+    fun increaseNumberOfAttempts() {
+        preferencesProvider.putInt(PREFERENCE_LOCK_ATTEMPTS, getNumberOfAttempts().plus(1))
+        preferencesProvider.putLong(PREFERENCE_LAST_UNLOCK_ATTEMPT_TIMESTAMP, SystemClock.elapsedRealtime())
+    }
+
+    fun resetNumberOfAttempts() =
+        preferencesProvider.putInt(PREFERENCE_LOCK_ATTEMPTS, 0)
+
+    fun getTimeToUnlockLeft(): Long {
+        val timeLocked = 1.5.pow(getNumberOfAttempts()).toLong().times(1000)
+        val lastUnlockAttempt = preferencesProvider.getLong(PREFERENCE_LAST_UNLOCK_ATTEMPT_TIMESTAMP, 0)
+        return max(0, (lastUnlockAttempt + timeLocked) - SystemClock.elapsedRealtime())
+    }
+
+    fun initUnlockTimer() {
+        object : CountDownTimer(getTimeToUnlockLeft(), 1000) {
+            override fun onTick(millisUntilFinished: Long) {
+                val minutes = TimeUnit.MINUTES.convert(millisUntilFinished.plus(1000), TimeUnit.MILLISECONDS)
+                val seconds = TimeUnit.SECONDS.convert(millisUntilFinished.plus(1000), TimeUnit.MILLISECONDS).rem(60)
+                val timeString = String.format("%02d:%02d", minutes, seconds)
```

Comment:
Are we only interested in minutes and seconds? What happens when the time is over an hour?

Reply:
Good catch, I'll add a field for the hours in case the delay is greater than 60 minutes, so the format remains as `hh:mm:ss` 👍